### ISSUE TITLE
ContractService and Policy Integration

### ISSUE THREAD

**Issue by jimmarino** [eclipse-dataspaceconnector/DataSpaceConnector#227]
# Overview

We need to support policy-based data contracts in the runtime. This issue will outline the work required to achieve this functionality in the EDC.

Policies are used for two purposes:

1. To define access rights for 1..N `Assets`
2. To define usage requirements for 1..N `Assets`

For example, to restrict an asset to "gold partners" and require users to store the asset data in a geographic region, two policy `Permission`s may be used:

- Permission: Partner=GOLD 
- Permission: Region=EU 

A policy associated with an asset is evaluated against the `claims` (e.g. verifiable credentials) and `attributes` of a `ParticipantAgent`. If the claims and attributes
satisfy the policy, a `ParticipantAgent` will be granted access to the asset.

# Participant Agent

A `ParticipantAgent` is a connector system running on behalf of a dataspace `Participant`. A `ParticipantAgent` is not the same as a dataspace `Participant` since the former may have claims and attributes that are distinct from the latter. For example, the Acme organization may have connector systems deployed in different geographical regions. 
While Acme may possess the claim/attribute Partner=GOLD, only the connector running in the EU will have the claim Region=EU. 

A `ParticipantAgent` is constructed by the `ParticipantAgentService` based on the `ClaimsToken` associated with the request.

### Claims and Attributes

When a `ParticipantAgent` is constructed, its claims derive from verifiable credentials; attributes are key/value pairs that are decorated by extension(s) in the provider runtime. For example, Partner=GOLD may be an attribute added to the `ParticipantAgent` by an extension that performs a lookup in a back-end system using the identity associated with the 
`ClaimsToken`.
 
# Contract Offers

A `ContractOffer` defines the policy that is _attached_ to a set of assets. The relationship between a `ContractOffer` and an asset is a multiplicity: a `ContractOffer` may be attached to multiple assets and an asset may have multiple `ContractOffers` attached to it. Policy attachment is the process of associating a policy with 1..N assets; this is performed using an `AssetSelectorExpression`. The latter are serializable representations of an asset grouping.

### ContractOfferFramework and ContractDefinition

A `ContractOfferFramework` is an extensible service that returns the set of `ContractDefinitions`s a `ParticipantAgent` qualifies for. In business terms, a `ContractDefinition` 
is the terms of use associated with one or more assets a `ParticipantAgent` has access to. It may be based on an existing business relationship between the data provider and
data requester, or it may be ad-hoc terms constructed by the data provider based on automated rules. 

A `ContractDefinition` contains one `Policy` that encapsulates the _usage requirements_ and an `AssetSelectorExpression`.
                                     
### ContractOfferService

The `ContractOfferService` receives a `ContractOfferQuery` from a client and delegates to the `ParticipantAgentService` and then requests the set of `ContractDefinitions` for a 
`ParticipantAgent` from the `ContractOfferFramework`. It takes the results and queries the `AssetIndex` for matching assets. Each `ContractDefinition` is combined with the returned
assets to construct contract offers.

### Query Criterion

The `ContractOfferQuery` may contain a collection of `Criterion` that narrow the scope of assets a client connector is interested in. For example, a client may query assets belonging to a particular category or a specific set of assets using their ids. The `Criterion` is passed to the `AssetIndex` along with the `AssetSelector`.

### ContractOfferFramework Implementations

The default ContractOfferFramework implementation will return an empty set of `ContractOffers`. This implementation can be replaced by a custom one or the sample implementation.

The sample `ContractOfferFramework` implementation will offer more capabilities. It will take a set of `ContractDescriptors` (this is just a working term; the name is not finalized)
that will contain two `Policy` instances, access policy and usage policy. The reason for the separation is that access policy should not be advertised to clients 
(e.g. Partner = GOLD or Partner = PLATINUM). The implementation will invoke the `Policy Engine`, passing it the union of the access and usage policies for evaluation against the 
`ParticipantAgent`. When a policy union is satisfied, the associated `ContractDescriptor` will be returned as a `ContractDefinition`. The `ContractDefinition` will only contain
the usage policies.

It is intended that the sample `ContractOfferFramework` use the `DataLoader` to register `ContractDescriptors`. The `DataLoader` will be described in a separate issue.

### PolicyEngine

The `PolicyEngine` is a runtime service that performs policy evaluation. Policy functions contributed to the runtime as extensions can be registered with the engine. 
These functions will be invoked at specific points in the evaluation process. For example, determining if a `ParticipantAgent` has an attribute Partner=GOLD. Policy functions 
can act as a `Policy Information Point` (PIP) common to many policy systems.  

The `PolicyEngine` will construct and use a `PolicyEvaluator`. 

# Catalog Query Engine

The catalog `QueryEngine` will use the `ContractOfferService` infrastructure.

# PR Descriptions

### Part 1: Refactor ContractService and ContractOfferFramework interfaces

The `ContractOfferService` and ContractOfferFramework interfaces will be refactored to implement the approach detailed in this document. This will involve removal of the 
`ContractTemplate` and reorganization of functionality by placing it in the `ContractService`. This refactoring will decouple the `ContractOfferFramework` from assets 
(the latter will only refer to AssetSelectorExpressions). Criterion will also be added to the input for the `ContractOfferService` but not passed to the AssetIndex.

### Part 2: Policy Engine implementation and working ContractDescriptors

The PolicyEngine will be implemented and integrated into the sample ContractOfferFramework. A mechanism for loading `ContractDescriptors` will also be provided, along with 
test descriptors that match on publicly available data and Partner=GOLD.

### Part 3: Catalog and AssetIndex integration
                                                                  
The `AssetIndex` will be updated to take additional Criterion passed to the `ContractOfferService`. Since Criterion is now a general type, it will be moved from the `asset` package
to its own package.

The Catalog subsystem (`QueryEngineImpl`) will be integrated with the `ContractService`.

### Related Work Items: Refactor Criterion

In a future PR, we may want to refactor the `Criterion` type to allow for _OR_ conditions. This will require the introduction of an expression concept that can group multiple
`Criterion`. Note the goal should *NOT* be to create a generalized query language as that can lead to an attack vector in the runtime. At most, simple _AND_ and _OR_ conditions
should be supported.

### Related Work Items: DataLoader

We will require a generalized framework for loading data into a connector system. This loader must support multiple data types through extensibility (e.g. assets and their associated data, contract descriptors, etc.), high-volume batch processing, and multiple ingress methods. For example, the loader may be invoked from the command line and fed
a file, from a REST API, or as an embedded SDK from a messaging system sink.

The loader should be provided in a set of separate PRs.

**Comment by DominikPinsel** [comment]
Good ideas 👍 

I probably haven't understand everything yet. Maybe it's the best idea if a create a separate comment for each assumption/question.

You wrote that we need to manage to kinds of identities. The `ParticipantAgent` and the `Participant`. Is it correct to assume that these two things are created by some parent component (e.g. an IDS controller) and den passed down using the `ContractOfferQuery`?

**Comment by DominikPinsel** [comment]
I assume the `ContractOfferService` would send all queries directly through to the `ParticipantAgentService`, without adding any value. Maybe we can move the functionality of the described `ParticipantAgentService` up to the `ContractOfferService`?

**Comment by DominikPinsel** [comment]
Is it correct to assume the `ParticipantAgent` and `Participant` are classes, that inherit from `Principal`?

**Comment by DominikPinsel** [comment]
@jimmarino Would it be fine if I start with the refactoring of the `ContractOfferFramework`? I would include
- `ContractOfferFramework` interface updates
- `ContractDefinition` creation
- `ContractDescriptors` creation
- `SampleContractOffer` extension update

**Comment by jimmarino** [comment]
Thanks for the good feedback @DominikPinsel. Here are some comments...

>You wrote about two kinds of identities: the ParticipantAgent and the Participant. Is it correct to assume that these two things are created by some parent component (e.g. an IDS controller) and den passed down using the ContractOfferQuery? 

Not quite an IDS component. There would be an EDC core concept of a ParticipantAgent and a ParticipantAgenService so it is a separate component, just not at the IDS level. 

Participant will have no materialization in the runtime, it's just a concept in the model (a Connector never interacts with a Participant directly, it interacts with the ParticipantAgent. 

> Is it correct to assume the `ParticipantAgent` and `Participant` are classes, that inherit from `java.security.Principal`?

java.security.Principal is not really relevant in this case as we are not trying to plug into a JAAS based-system. Principal does not provide any useful methods in our case, so ParticipantAgent will not implement that interface.

>I assume the ContractOfferService would send all queries directly through to the ParticipantAgentService, without adding any value. Maybe we can move the functionality of the described ParticipantAgentService up to the ContractOfferService?

ContractOfferService is really just a coordinator between the ParticipantAgentService, ContractOfferFrameworkService, and the AssetIndex. I think each of those three services is distinct from the ContractOfferService. Let's see how the code evolves.

>@jimmarino Would it be fine if I start with the refactoring of the ContractOfferFramework? I would include...

@DominikPinsel let's work on this together. I'll start with a pull request tomorrow. Can you commit to an external repository, in which case I can set one up? If not, we'll need to come up with a way for you to contribute, which we can discuss tomorrow.

**Comment by DominikPinsel** [comment]
> > @jimmarino Would it be fine if I start with the refactoring of the ContractOfferFramework? I would include...
> 
> @DominikPinsel let's work on this together. I'll start with a pull request tomorrow. Can you commit to an external repository, in which case I can set one up? If not, we'll need to come up with a way for you to contribute, which we can discuss tomorrow.

It's also fine if you do it. I removed the comment already, because when I wrote it I wasn't aware that you intent to contribute all the code. Therefore I thought it might be a good idea to contribute the part,  some issues I created yesterday build upon. But it's also perfectly fine for me if you do it 👍

**Comment by DominikPinsel** [comment]
> > You wrote about two kinds of identities: the ParticipantAgent and the Participant. Is it correct to assume that these two things are created by some parent component (e.g. an IDS controller) and den passed down using the ContractOfferQuery?
> 
> Not quite an IDS component. There would be an EDC core concept of a ParticipantAgent and a ParticipantAgenService so it is a separate component, just not at the IDS level.
> 
> Participant will have no materialization in the runtime, it's just a concept in the model (a connector never interacts with a Participant directly, it interacts with the ParticipantAgent).

Sounds interesting :) I think I haven't understand it yet. But maybe I will just wait for the first code then.

**Comment by DominikPinsel** [comment]
> > I assume the ContractOfferService would send all queries directly through to the ParticipantAgentService, without adding any value. Maybe we can move the functionality of the described ParticipantAgentService up to the ContractOfferService?
> 
> ContractOfferService is really just a coordinator between the ParticipantAgentService, ContractOfferFrameworkService, and the AssetIndex. I think each of those three services is distinct from the ContractOfferService. Let's see how the code evolves.

Ah ok. This sounds a lot more logical. This is what I sketched yesterday when I read the text the first time. Looks like I need to update it once more 😄 

<img width="1155" alt="image" src="https://user-images.githubusercontent.com/20866918/141426030-3b0dbc14-5e89-4c97-91ea-10a53abf5dda.png">

**Comment by jimmarino** [comment]
> > > @jimmarino Would it be fine if I start with the refactoring of the ContractOfferFramework? I would include...
> > 
> > 
> > @DominikPinsel let's work on this together. I'll start with a pull request tomorrow. Can you commit to an external repository, in which case I can set one up? If not, we'll need to come up with a way for you to contribute, which we can discuss tomorrow.
> 
> It's also fine if you do it. I removed the comment already, because when I wrote it I wasn't aware that you intent to contribute all the code. Therefore I thought it might be a good idea to contribute the part, some issues I created yesterday build upon. But it's also perfectly fine for me if you do it 👍

@DominikPinsel actually, I'd prefer we worked on this together. Is there a way for you to be able to commit directly to a repo I create? If not, let's think of another way to do it.

**Comment by jimmarino** [comment]
> > > I assume the ContractOfferService would send all queries directly through to the ParticipantAgentService, without adding any value. Maybe we can move the functionality of the described ParticipantAgentService up to the ContractOfferService?
> > 
> > 
> > ContractOfferService is really just a coordinator between the ParticipantAgentService, ContractOfferFrameworkService, and the AssetIndex. I think each of those three services is distinct from the ContractOfferService. Let's see how the code evolves.
> 
> Ah ok. This sounds a lot more logical. This is what I sketched yesterday when I read the text the first time. Looks like I need to update it once more 😄
> 
> <img alt="image" width="1155" src="https://user-images.githubusercontent.com/20866918/141426030-3b0dbc14-5e89-4c97-91ea-10a53abf5dda.png">

I like the picture. One thing is the PAS does not have lines with the COF or the AssetIndex; only the ContractOfferService has dependencies on the PAS and COF. 

The PAS should be a really simply service. It takes a ClaimToken and outputs a ParticipantAgent. Extensions can register ParticipantAgentServiceExtensions that take a ClaimToken and return a Map of attributes that are added to the ParticipantAgent. Once that happens, the PA is sent to the COF for evaluation.

Let's talk about the COF and PolicyEngine next. The COF is going to need a PolicyEngine. I think the PE needs to be a core service that other extensions can register policy functions with. Initially, we had put in a IdsPolicyServiceImpl but that can be removed and replaced with the more generalized PolicyEngine. The mechanism should, however, remain consistent. Do you want to take a look at how that works and then we can discuss the PolicyEngine?

**Comment by DominikPinsel** [comment]
> The mechanism should, however, remain consistent. Do you want to take a look at how that works and then we can discuss the PolicyEngine?

Sure 👍 I will try to sketch something we can use as starting point for a discussion

**Comment by MoritzKeppler** [comment]
> 
> 
> I assume you suggested that the `ContractDefinition` contains an `AsselSelectorExpression` and a `Policy (usage)`. Maybe we can add another field here that indicates whether the same contract should be generated for each asset, or whether it should be one large contract for all assets.

my question goes into the same direction: A `ContractDefinition` results in one corresponding `ContractOffer` - just containing a fixed list of Assets instead of an AssetSelector. In parallel each `Policy` contains a target that must be filled with a reference to an Asset (and later in the IDS layer with a reference to an artifact, representation, resource etc.).
Right now the target does not get filled at all (right?) and we create just contracts referencing all assets at once. 
I expected that we create a ContractOffer for every Asset that matches the selector of a ContractDefinition.

**Comment by jimmarino** [comment]
Implemented
